all: instrument with OpenCensus

odeke-em opened the pull request:
Started instrumentation for monitoring and tracing with OpenCensus

odeke-em committed:
all: instrument with OpenCensus

Started instrumentation for monitoring and tracing with OpenCensus

odeke-em committed:
src/main/java/redis/clients/jedis: instrument Jedis methods

Instrumented all Jedis->redis methods by tracing

odeke-em committed:
src/main/java/redis/clients/jedis: roundtrip time tracking span

Introduce a roundtrip time tracking span that on end will
both end the span but also record stats for the overall
roundtrip time for any command and then tag the metric with
the command name.

odeke-em committed:
all: use milliseconds for latency + coerce maven dependency Guava

* Ensure that we coerce/pin a Guava dependency in the maven file
as per
https://github.com/census-instrumentation/opencensus-java/issues/1220#issuecomment-393340392

* Use milliseconds instead of seconds for latency measurements
* Switch out all BinaryJedis.java usages of
    Observability.startSpan(spanName)
to
    Observability.createRoundtripTrackingSpan(spanName, metricTagName);

odeke-em commented:
And with that we now have

### Metrics
![screen shot 2018-05-30 at 4 28 22 pm](https://user-images.githubusercontent.com/4898263/40752979-ceab4774-6426-11e8-88bf-4fe71cbe302c.png)

![screen shot 2018-05-30 at 4 31 35 pm](https://user-images.githubusercontent.com/4898263/40753086-403a14a6-6427-11e8-8d73-a47bb2f4ca56.png)


### Tracing
![screen shot 2018-05-30 at 4 27 38 pm](https://user-images.githubusercontent.com/4898263/40752982-d5e158ee-6426-11e8-8f80-c84d744e508a.png)

songy23 commented on src/main/java/redis/clients/jedis/Observability.java:
```diff
@@ -0,0 +1,259 @@
+                    noKeys)};
+
+
+        ViewManager vmgr = Stats.getViewManager();
```
nit: `ViewManager` could be `private static final`

odeke-em replied:
Thanks @songy23, I'll fix that.

odeke-em replied:
Oh wait, it is only used in that function and in about only 3 lines, so perhaps not much of a worry?

songy23 replied:
Yes in this demo code it's fine to leave as-is. For common use cases, we usually need to keep a reference to ViewManager (as a static variable). For example the [quick start](https://github.com/census-instrumentation/opencensus-java/blob/master/examples/src/main/java/io/opencensus/examples/helloworld/QuickStart.java#L51).

odeke-em replied:
This isn't demo code though :) This is the actual instrument client code that will be used by everyone. I thought about making a static client yesterday but this function only gets called once and that variable is used only down here below so didn't necessitate that.

songy23 replied:
SGTM

songy23 reviewed (commented):
Stats and tracing look good to me, though I'm not quite familiar on how this can be used in redis.

odeke-em committed:
Connection-centric monitoring

* Tracking the number of opened and closed connections
* Tracking the number of errors when dealing with closed/closing
connections
* Tracking the number of read routine invocations, bytes read and read errors
* Observability.record*Stat helpers for ints to avoid the current
clutter/ugliness of using `new Long` in userland

odeke-em committed:
Observability: use ConnectionsReused; JAR instructions

* Ensure that we track the number of connections reused
i.e. the number of times we are already connected and don't
have to reconnect to the server
* Provide some instructions in the README on how to generate
the JAR and install it locally
* Remove repeated annotation "Connecting" in favor of prior and more
comprehensive annotation
* Fix a nit about using `new Long(1)`, to use the nicer
signature for `Observability.recordStat(MeasureLong, int)`

odeke-em committed:
Observability: use MConnectionErrors for connection errors

Since all IOExceptions from the read or write end of a socket
relate to connections, instead of MConnectionClosedErrors, group
them all to MConnectionErrors. However, add a tag "phase"
to help give more context as to which phase is occuring e.g.
* connect
* disconnect
* flush
* read
* readProtocolWithCheckingBroken
* write

Also for JedisDataExceptions, record MReadErrors.

odeke-em committed:
Observability: use tracer.startScopedSpan and retain the scope

For any started span, we need to always ensure that it is
inserted into the current scope first lest buggy library
uses will show created spans as not existing within the same
scope.

odeke-em commented on src/main/java/redis/clients/jedis/Observability.java:
```diff
@@ -107,12 +112,14 @@ public static void annotateSpan(Span span, String description, Attribute ...attr
         public RoundtripTrackingSpan(String name, String commandName) {
-            this.span = tracer.spanBuilder(name).startSpan();
+            this.spanScope = tracer.spanBuilder(name).startScopedSpan();
+            this.span = tracer.getCurrentSpan();
```
@songy23 this is how I've done it

songy23 replied:
You also need to close the scope after setting it, something like:
```java
spanScope.close();
span.end();
```
So it might be easier to just return the span, and set span into scope in the code that calls RoundtripTrackingSpan() instead of doing it in RoundtripTrackingSpan().

odeke-em commented on src/main/java/redis/clients/jedis/Observability.java:
```diff
@@ -93,8 +93,13 @@ public static void recordTaggedStat(TagKey key, String value, MeasureDouble md,
     public static Span startSpan(String name) {
-        return tracer.spanBuilder(name).startSpan();
+        tracer.spanBuilder(name).startScopedSpan();
+        return tracer.getCurrentSpan();
```
@songy23 ditto here

songy23 replied:
Same here, you'll need to keep a reference of scoped span in order to close it.

odeke-em committed:
Observability: create a ScopedSpan

For a span to be properly ended, its scope
has to be ended as well hence provide a class
that has both the scope and span; invoking .end
on it will end both the scope and span.

odeke-em committed:
observability: consolidate errors to use single measure but tagkeys

Consolidate the various errors to use one single measure `MErrors`
However, provide various tag keys to discern them e.g.
* command -- the name of the command e.g. "getMany", "read", "ttl"
* phase   -- the part of the command e.g. "read", "flush"
* enum    -- an enumeration to distinguish a part e.g. "has_prefix"

odeke-em committed:
observability: use KeyCommandName tag appropriately

For views for the following measures:
* MBytesWritten
* MRoundtripLatencyMilliseconds
* MWrites
